### PR TITLE
Refactor Validating Webhooks

### PR DESCRIPTION
**Pull request by davewalter**
## Is there a related GitHub Issue?
#1174

## What is this change about?
This PR refactors the validating webhooks to use the pattern described in [this article](https://betterprogramming.pub/writing-custom-kubernetes-controller-and-webhooks-141230820e9) instead of registering custom handlers.

## Does this PR introduce a breaking change?
No

## Acceptance Steps
Tests pass and validation rules are still applied.

## Tag your pair, your PM, and/or team
@Birdrock

### CODE REVIEW

**Comment by davewalter** [comment]
ToDo:
* Refactor CFSpace validating webhook (blocked on #1167)
* Refactor similar validating webhooks (the ones that make use of a duplicate validator) to avoid code duplication. For example, the code that converts the error returned by the duplicate validator could be moved into the duplicate validator itself.

File: controllers/webhooks/networking/cfroute_validation.go
Comment (kieron-dev) [inline controllers/webhooks/networking/cfroute_validation.go:95]: It might improve readability to extract the duplicateValidator call from the method arg here. Then I would also be inclined to do the nil check here (even though that would be repeated in the update and delete validation methods) and remove the nil check from the generateDuplicateValidationError method to make it more focused.

**Review by kieron-dev — COMMENTED** [review]
Overall, this is a really nice change. It's great to lose the code checking whether to decode the old or new objects based on the operation type.

When looking at this with @danail-branekov, we talked about whether it would be nice to separate the different validations into composable types. It might be good, but I guess that's not really the object of this PR though and would require a bit of framework code.

Should the *_validation files be renamed *_validator to match the types btw?

File: controllers/webhooks/networking/cfroute_validation.go
Comment (davewalter) [reply controllers/webhooks/networking/cfroute_validation.go:95]: I was contemplating moving the logic from `generateDuplicateValidationError` into the `duplicateValidator.Validate*` methods as there seemed to be a lot of duplication of handling of the error across all the webhooks.

**Comment by davewalter** [comment]
Thanks for taking a look at this @kieron-dev and @danail-branekov.

> When looking at this with @danail-branekov, we talked about whether it would be nice to separate the different validations into composable types. It might be good, but I guess that's not really the object of this PR though and would require a bit of framework code.

I definitely think that this has helped to surface some potential refactors such as this, but I agree that it would be better to create follow-on chores for them, rather than delay getting this merged.

> Should the *_validation files be renamed *_validator to match the types btw?

Definitely. I'll add it to the list of items to complete before merging this.

**Comment by danail-branekov** [comment]
Note that we are about to introduce a `CFTask` validator with PR https://github.com/cloudfoundry/korifi/pull/1200, could you please consider it as well when rebasing the PR?

**Comment by davewalter** [comment]
> BTW, it is worth mentioning that the [kubebuilder book](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html) suggests that validation is implemented on top of the crd type itself, rather than having a separate validator type. This has the benefit that the validators would not need to cast the generic runtime object to a specific type before actually performing the validation.

Most (if not all) of the validating webhooks make use of K8s clients, duplicate validators, and placement validators to check for other resources on the cluster. If we implement the validators on top of the CRD types, we would have to add these things to the type structs, which would pollute them. This felt like a cleaner solution.

**Review by matt-royal — APPROVED** [review]
Dave and I discussed the PR on Zoom. I made some suggestions for refactoring the json encoding and decoding of errors, but otherwise this looks good to me.

**Pull request merged**
